pkg/config/cfgvars.go

```diff
@@ -0,0 +1,303 @@
+		EtcdDataDir:                filepath.Join(dataDir, "etcd"),
+		KineSocketPath:             filepath.Join(runDir, constant.KineSocket),
+		KonnectivitySocketDir:      filepath.Join(runDir, "konnectivity-server"),
+		KubeletAuthConfigPath:      filepath.Join(dataDir, "kubelet.conf"),
+		KubeletVolumePluginDir:     constant.KubeletVolumePluginDir,
+		ManifestsDir:               filepath.Join(dataDir, "manifests"),
+		RunDir:                     runDir,
+		KonnectivityKubeConfigPath: filepath.Join(certDir, "konnectivity.conf"),
+		RuntimeConfigPath:          filepath.Join(runDir, "k0s.yaml"),
+		StatusSocketPath:           filepath.Join(runDir, "status.sock"),
+		StartupConfigPath:          constant.K0sConfigPathDefault,
+
+		// Helm Config
+		HelmHome:             helmHome,
+		HelmRepositoryCache:  filepath.Join(helmHome, "cache"),
+		HelmRepositoryConfig: filepath.Join(helmHome, "repositories.yaml"),
+
+		origin: CfgVarsOriginDefault,
+	}
+
+	if cmd != nil {
+		WithCommand(cmd)(vars)
+	}
+
+	return vars, nil
+}
+
+func (c *CfgVars) Cleanup() error {
+	if c.origin == CfgVarsOriginDefault && file.Exists(c.RuntimeConfigPath) {
+		return os.Remove(c.RuntimeConfigPath)
+	}
+	return nil
+}
+
+func (c *CfgVars) defaultStorageSpec() *v1beta1.StorageSpec {
+	if c.DefaultStorageType == v1beta1.KineStorageType {
+		return &v1beta1.StorageSpec{
+			Type: v1beta1.KineStorageType,
+			Kine: v1beta1.DefaultKineConfig(c.DataDir),
+		}
+	}
+
+	return v1beta1.DefaultStorageSpec()
+}
+
+func (c *CfgVars) NodeConfig() (*v1beta1.ClusterConfig, error) {
+	if c.nodeConfig != nil {
+		return c.nodeConfig, nil
+	}
+
+	if c.origin == CfgVarsOriginRuntime {
+		return nil, fmt.Errorf("runtime config is not available")
+	}
+
+	if c.StartupConfigPath == "" {
+		return nil, fmt.Errorf("config path is not set")
+	}
+
+	var nodeConfig *v1beta1.ClusterConfig
+
+	cfgContent, err := os.ReadFile(c.StartupConfigPath)
+	if errors.Is(err, os.ErrNotExist) && c.StartupConfigPath == constant.K0sConfigPathDefault {
+		nodeConfig = v1beta1.DefaultClusterConfig(c.defaultStorageSpec())
+	} else if err == nil {
+		cfg, err := v1beta1.ConfigFromString(string(cfgContent), c.defaultStorageSpec())
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig = cfg
+	} else {
+		return nil, err
+	}
+
+	if nodeConfig.Spec.Storage.Type == v1beta1.KineStorageType && nodeConfig.Spec.Storage.Kine == nil {
+		nodeConfig.Spec.Storage.Kine = v1beta1.DefaultKineConfig(c.DataDir)
+	}
+
+	c.nodeConfig = nodeConfig
+
+	return nodeConfig, nil
+}
+
+func (c *CfgVars) ClusterConfig(ctx context.Context, kubeClientFactory kubeutil.ClientFactoryInterface) (*v1beta1.ClusterConfig, error) {
+	if !c.EnableDynamicConfig {
+		return c.NodeConfig()
+	}
+
+	var apiConfig *v1beta1.ClusterConfig
+
+	client, err := kubeClientFactory.GetConfigClient()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := retry.Do(
+		func() (err error) {
+			ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
```

Comment:
I think the old timeout was ten seconds. I'd at least restore that. Maybe remove the timeout completely and rely on the client-go timeout settings instead?

Reply:
The only things calling `ClusterConfig()` are 1) controller startup 2) airgap list-images. 

I think the controller could live without it. Components that require the "dynamic config" could probably just wait for config reconciliation before starting up. It will fail every time when there are no other controllers in the cluster. 

I don't know if even 10 seconds is enough for list-images if you run `sudo k0s start` and `k0s airgap list-images` right after it.  Maybe it should instead wait for status socket before proceeding.

Reply:
Does list-images need to work with dynamic config?

Reply:
The config business in controller node/clustercomponents is complicated.

### node components:

- all of the nodeComponents that need clusterConfig receive it during construct, ie. &foo.Foo{Config: config} or foo.NewFoo(config)
- none of the nodeComponents have Reconcile, they will never receive a config in any other way after that

### cluster components:

- some of the clusterComponents receive a nodeconfig during constructor, some don't.
- many of them have a Reconcile function, so they will receive a new config when:
  * ClusterConfigReconcilier first starts
  * When ClusterConfigReconcilier detects a dynamic config change
  
 The startup process is somewhat exotic:
  
  1. All of the components are `Add()`ed to cluster component manager
  2. Finally component manager's `Start()` is called
  3. Start iterates over all of the `Add()`ed components and calls `Start()` for each and then waits for `Ready()` if the component has defined that function.
  4. When the iteration encounters the `ClusterConfigReconcilier` component somewhere middle of the way, it calls the `Start()` on it like for any other component, which makes ClusterConfigReconciler call component manager's `Reconcile(ctx, cfg)`
  5. Component manager's `Reconcile` iterates over all of the `Add`ed components (we're still in the middle of the iterator in step 3) and calls `Reconcile(ctx, cfg)` on them if they have defined this function.
  6. At this point, all of the components that have a Reconcile function will receive a new config (local or dynamic) 
  7. The iteration in step 3 continues to `Start()` the rest of the components
  
  So,

- the components before step 4 are started with either the local config or no config
- the components after step 4 receive a new config via `Reconcile()` (if they implement it), before they are started

  This needs to be cleaned up somehow.

Reply:
Some additional issues:

#### The component manager:

- It has an `Init()` function that is called once for both of the managers in the controller.
- It also has the `Add()` function. The function tries to call `Reconcile` on the components in case the manager has a config available. It never has unless there's something calling the manager's `Reconcile`,  which only happens once it has been started.
- Any component added after the `Init()` call will not be initialized when `Add()`ed.
- The `Start()` function iterates over all of the components to call `Start` on them and then returns. This happens once.
- Nothing starts or initializes any components that are added after that.
- Nothing restarts components that fail
- If a component fails, nothing kills the manager
- the "reconcile" term is used rather sloppily for multiple things and it's slightly difficult to figure out what it means for each of the components. I think the dynamic config reconciliation related functions should be named uniquely.

#### Regarding the components:

- There are multiple components that have just `return nil` in their `Init()` and `Start()` and the actual functionality is in `Reconcile()`.
- Some components have functionality in their `Start` but they are one-shot, the component "starts" but nothing is actually left running.
- There are several components that are added to the `clusterComponents` manager that do not have a `Reconcile` function, so they could just as well be in the `nodeComponents` manager that doesn't reconcile.
- The `pkg/component/controller` package defines 48 structs and 5 interfaces. The "node components" and "cluster components" are all mixed up in there in a single directory without any separation.
- Many of them have `// Run ....` doc comment on top of the `Start()` function, a remnant from some function rename I suppose

#### The prober:

- The only components that implement the prober healthcheck interface are: Konnectivity and `mockComponent`.
- The only components that implement the prober event emitter interface are: Konnectivity, OCIBundleReconcilier (which is one-shot, `Start` just unpacks stuff and returns) and `mockComponentWithEvents`.

#### The cluster config reconciler:

- Nothing restarts the watcher if the server or a watchtimeout closes it
- It does not differentiate between add, create, update, delete
- It does not check for nils. If it receives a *v1beta1.ClusterConfig typed nil, it will call `Validate()` on that, which will return `nil` and then it proceeds to reconcile using the nil.
- Nothing tells you if the component has actually stopped, not even the experimental `k0s status components` subcommand. It doesn't implement event emitter or health check interfaces. the only way you notice is that config reconciliation does not happen.

Reply:
The components are started and reconcilied sequentially, not sure if it's necessary.

Reply:
> The only things calling `ClusterConfig()` are 1) controller startup 2) airgap list-images.

Now only number 2.

Does "airgap list-images" need to work with dynamic config, @jnummelin ?

Reply:
Now airgap list-images will error out if dynamic is enabled. Now nothing uses FetchDynamicConfig() so it has been removed. Now configs are never merged.